design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,341 @@
+    - potentially more complex
+
+An example pseudo-implementation of this could look like this:
+```rust
+fn skip_if_done<F>(func: F)
+where
+    F: Fn() + 'static,
+{
+    // Decorator logic here
+    // Check the state file and decide whether to execute the function or not
+
+    if should_execute {
+        func();
+        // write func outputs to state file
+    } else {
+        // Do nothing or return an appropriate value
+    }
+}
+
+fn main() {
+    skip_if_done(|| {
+        dispatcher.send_eth(...);
+    });
+}
+```
+
+- Use a procedural macro to automatically apply decorator-like behaviour to functions, which will decide on whether
+given function should be executetd or not (based on state file)
+  - pros:
+    - elegant solution
+    - could be applied selectively (per function) or globally (to main)
+  - cons:
+    - more complex solution
+    - probably not very suitable for runtime based behaviours (like reading json file) - requires more research
+
+An example pseudo-implementation of this could look like this:
+```rust
+// skip_if_done.rs
+use proc_macro_hack::proc_macro_hack;
+
+#[proc_macro_hack]
+pub use skip_if_done_impl::skip_if_done;
+
+
+// skip_if_done_impl.rs
+use quote::quote;
+use std::env;
+
+fn should_execute() -> bool {
+    // Your logic to check whether the function should be executed goes here
+    true
+}
+
+#[macro_export]
+macro_rules! skip_if_done {
+    ($($tokens:tt)*) => {
+        {
+            if should_execute() {
+                $($tokens)*
+            }
+        }
+    };
+}
+
+// main.rs
+#[path = "skip_if_done.rs"]
+mod skip_if_done;
+
+fn main() {
+    skip_if_done! {
+        dispatcher.send_eth(...);
+    }
+}
+```
+
+Given all this, the most obvious solutions seem to be:
+  - for cast subcommands functions:
+    - adjust them to handle checking the file
+  - for dispatcher functions:
+    - implement a higher order function
+
+Picking those would ensure us that:
+  - syntax is relatively boiler-code free (only dispatcher functions are annotated)
+  - idempotency is achieved
+  - it is quite easy to write a script
+
+### example script
+An example deployment script could look like this:
+
+```cairo
+// we might need to rename account functions to avoid confusion
+use cast::starknet_commands::account::create::create as create_account
+use cast::starknet_commands::account::deploy::deploy as deploy_account
+use cast::{get_provider, get_account};
+use cast::starknet_commands::{declare, deploy, invoke, call};
+(...)
+
+fn make_account(provider: &JsonRpcClient<HttpTransport>) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
+  let mut prefunder = get_account("user", &provider)?; // the user that will prefund new account 
+  let user = create_account("user1", &provider);
+  let prefund_account = invoke("0x123", "deposit", [1234, user.address], &prefunder);
```

Comment:
Do we necessarily want to support funding from the script? If so, shouldn't it be with a dispatcher as well?

Reply:
This is not necessary, I just thought it was a low hanging fruit to show one can achieve this by simply invoking a contract. For first interation(s) I thought we would just require users to provide the dispatchers, so it would be doable but with additional input from users. would that make sense?

Reply:
Might be pushing it for an initial release, but we can have some kind of `abigen` to create super simple dispatchers